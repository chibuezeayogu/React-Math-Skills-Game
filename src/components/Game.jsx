import React, { Component, Fragment } from 'react';
import PropTypes, { bool } from 'prop-types';
import _ from 'lodash';

import Numbers from './Numbers';
import DoneFrame from './DoneFrame';
import Button from './Button';
import Answers from './Answers';
import Stars from './Stars';
import possibleCombinationSum from '../utils/possibleCombinationSum';


/**
 * 
 * @class Game
 * 
 * @extends Component
 */
class Game extends Component {
  static randomNumber = () => 1 + Math.floor(Math.random() * 9);
  
  static initialState = () => ({
    selectedNumbers: [],
    randomNumberOfStars: Game.randomNumber(),
    answerIsCorrect: null,
    usedNumbers: [],
    redraws: 5,
    doneStatus: null
  });
  
  constructor(props) {
    super(props);
    this.state = Game.initialState();
  }
  
  /**
   * 
   * @method resetGame
   * 
   * @memberof Game
   * 
   * @returns {void}
   */
  resetGame = () => this.setState(Game.initialState());
  
  /**
   * 
   * @method selectNumber
   * 
   * @memberof Game
   * 
   * @param {integer} clickedNumber
   * 
   * @returns {void}
   */
  selectNumber = (clickedNumber) => {
    if (this.state.selectedNumbers.indexOf(Number(clickedNumber)) >= 0) {
      return;
    }

    this.setState(preState => ({
      answerIsCorrect: null,
      selectedNumbers: preState.selectedNumbers.concat(Number(clickedNumber))
    }));
  };
  
  /**
   * 
   * @method unselectNumber
   * 
   * @memberof Game
   * 
   * @param {integer} clickedNumber
   * 
   * @returns {void}
   */
  unselectNumber = (clickedNumber) => {
    this.setState(preState => ({
      answerIsCorrect: null,
      selectedNumbers: preState.selectedNumbers.filter(
        number => number !== Number(clickedNumber)
      ) 
    }));
  };
  
  /**
   * 
   * @method checkAnswer
   * 
   * @memberof Game
   * 
   * @returns {void}
   */
  checkAnswer = () => {
    this.setState(prevState => ({
      answerIsCorrect: prevState.randomNumberOfStars === prevState.selectedNumbers.reduce((acc, n) => acc + n, 0)
    }));
  };
  
  /**
   * 
   * @method acceptAnswer
   * 
   * @memberof Game
   * 
   * @returns {void}
   */
  acceptAnswer = () => {
    this.setState(preState => ({
      usedNumbers: preState.usedNumbers.concat(preState.selectedNumbers),
      randomNumberOfStars: Game.randomNumber(),
      answerIsCorrect: null,
      selectedNumbers: []
    }), this.updateDoneStatus);
  };
  
  /**
   * 
   * @method redraw
   * 
   * @memberof Game
   * 
   * @returns {void}
   */
  redraw = () => {
    if (this.state.redraws === 0) { return; }
    this.setState(preState => ({
      randomNumberOfStars: Game.randomNumber(),
      answerIsCorrect: null,
      selectedNumbers: [],
      redraws: preState.redraws - 1
    }), this.updateDoneStatus);
  }; 

  /**
   * 
   * @method redraw
   * 
   * @memberof Game
   * 
   * @param {number} ramdomNumberOfStars
   * 
   * @param {array} sedNumbers
   * 
   * @returns {void}
   */
  possibleSolutions = ({ randomNumberOfStars, usedNumbers }) => {
    const possibleNumbers = _.range(1, 10).filter(
      number => usedNumbers.indexOf(number) === -1
    );
    
    return possibleCombinationSum(possibleNumbers, randomNumberOfStars);
  };
  
  /**
   * 
   * @method updateDoneStatus
   * 
   * @memberof Game
   * 
   * @returns {void}
   */
  updateDoneStatus = () => {
    this.setState(preState => {
      if (preState.usedNumbers.length === 9) {
        return { doneStatus: 'Done Nice!' };
      }
      if (preState.redraws === 0 && !this.possibleSolutions(preState)) {
        return { doneStatus: 'Game Over!' };
      }
    });
  };
  
  render() {
    const { 
      selectedNumbers,
      randomNumberOfStars,
      answerIsCorrect,
      usedNumbers,
      redraws,
      doneStatus
    } = this.state;
    const { progress } = this;
    return (
    <Fragment>
      <div className="jumbotron text-center">
        <h1>Math Skills Kids' Game</h1>
        <h2>Play Nine</h2>
      </div>
      <div className="container container-set">
        { doneStatus 
          ? <DoneFrame
              doneStatus={doneStatus}
              resetGame={this.resetGame}
            />
          : <div>
          <div className="row">
          <div className="card text-center justify-content-center col-sm-5">
            <Stars 
              numberOfStars={randomNumberOfStars} 
            />
          </div>
          <div className="text-center col-sm-2 col-md-2 col-lg-2 ">
            <Button
              selectedNumbers={selectedNumbers} 
              checkAnswer={this.checkAnswer}
              answerIsCorrect={answerIsCorrect}
              acceptAnswer={this.acceptAnswer}
              redraw={this.redraw}
              redraws={redraws}
            />
          </div>
          <div className="card text-center justify-content-center col-sm-5">
            <Answers 
              selectedNumbers={selectedNumbers} 
              unselectNumber={this.unselectNumber}
            />
          </div>
          </div>
          <br />
          <hr />
          <div className="card text-center justify-content-center col-sm-12" style={{ height: '100px' }}>
            <Numbers 
              selectedNumbers={selectedNumbers}
              selectNumber={this.selectNumber} 
              usedNumbers={usedNumbers}
            />
          </div>
          </div>
        }
        <br />
        <hr />
      </div>
    </Fragment>
    );
  }
}

export default Game;
