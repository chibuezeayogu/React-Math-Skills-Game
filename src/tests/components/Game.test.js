/* 
global jest 
expect 
*/
import React from 'react';
import { shallow } from 'enzyme';
import Game from '../../components/Game';

describe('Game Component', () => {
  const wrapper = shallow(<Game />);

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call resetGame method', () => {
    const resetGameSpy = jest.spyOn(wrapper.instance(), 'resetGame');
    wrapper.instance().resetGame();
    expect(resetGameSpy).toHaveBeenCalled();
  });
  it('should call selectNumber method', () => {
    const selectNumberSpy = jest.spyOn(wrapper.instance(), 'selectNumber');
    wrapper.instance().selectNumber(1);
    expect(selectNumberSpy).toHaveBeenCalled();
  });

  it('should call unselectNumber method', () => {
    const unselectNumberSpy = jest.spyOn(wrapper.instance(), 'unselectNumber');
    wrapper.instance().unselectNumber(1);
    expect(unselectNumberSpy).toHaveBeenCalled();
  });

  it('should call checkAnswer method', () => {
    const checkAnswerSpy = jest.spyOn(wrapper.instance(), 'checkAnswer');
    wrapper.instance().checkAnswer();
    expect(checkAnswerSpy).toHaveBeenCalled();
  });

  it('should call acceptAnswer method', () => {
    const acceptAnswerSpy = jest.spyOn(wrapper.instance(), 'acceptAnswer');
    wrapper.instance().acceptAnswer();
    expect(acceptAnswerSpy).toHaveBeenCalled();
  });

  it('should call redraw method', () => {
    const redrawSpy = jest.spyOn(wrapper.instance(), 'redraw');
    wrapper.instance().redraw();
    expect(redrawSpy).toHaveBeenCalled();
  });

  it('should call possibleSolutions method', () => {
    const spy = jest.spyOn(wrapper.instance(), 'possibleSolutions');
    wrapper.instance().possibleSolutions(
      { randomNumberOfStars: 2, usedNumbers: [2, 3] }
    );
    expect(spy).toHaveBeenCalled();
  });

  it('should call updateDoneStatus method', () => {
    const spy = jest.spyOn(wrapper.instance(), 'updateDoneStatus');
    wrapper.instance().updateDoneStatus();
    expect(spy).toHaveBeenCalled();
  });
});
