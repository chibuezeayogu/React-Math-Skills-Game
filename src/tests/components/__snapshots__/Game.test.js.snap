// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game Component should render correctly 1`] = `
<Fragment>
  <div
    className="container"
  >
    <ul
      className="nav justify-content-end"
    >
      <li
        className="nav-item"
      >
        <a
          className="nav-link active"
          data-target="#modal"
          data-toggle="modal"
        >
          How To Play
        </a>
      </li>
    </ul>
  </div>
  <Modal />
  <div
    className="jumbotron text-center"
  >
    <h1>
      Math Skills Kids' Game
    </h1>
    <h2>
      Play Nine
    </h2>
  </div>
  <div
    className="container container-set"
  >
    <div>
      <div
        className="row"
      >
        <div
          className="card text-center justify-content-center col-sm-5"
        >
          <Stars
            numberOfStars={3}
          />
        </div>
        <div
          className="text-center col-sm-2 col-md-2 col-lg-2 "
        >
          <Button
            acceptAnswer={[Function]}
            answerIsCorrect={null}
            checkAnswer={[Function]}
            redraw={[Function]}
            redraws={5}
            selectedNumbers={Array []}
          />
        </div>
        <div
          className="card text-center justify-content-center col-sm-5"
        >
          <Answers
            selectedNumbers={Array []}
            unselectNumber={[Function]}
          />
        </div>
      </div>
      <br />
      <hr />
      <div
        className="card text-center justify-content-center col-sm-12"
        style={
          Object {
            "height": "100px",
          }
        }
      >
        <Numbers
          selectNumber={[Function]}
          selectedNumbers={Array []}
          usedNumbers={Array []}
        />
      </div>
    </div>
    <br />
    <hr />
  </div>
</Fragment>
`;
